p2p/p2p.md

```diff
@@ -0,0 +1,220 @@
+`Swarm`'s responsibility is to receive and send network messages and connections using events propagated from `Net`, handle them - check that they are in a valid format and time and pass them over to the right protocol handler using `Demuxer`.
+
+*TODO*: Describe the patterns used in `Swarm` the event loop, state management and how to prevent deadlocks.
+Also about the bootstrap loop and the network loop.
+
+### P2P Protocols :
+  - `HandshakeProtocol` - Used to initiate a secured connection with peers
+  ```protobuf
+  // The initiator creates a HandshakeData object
+  // with a random IV based on the requested node Public Key
+  // and our own Private Key.
+  // The request also contains basic details
+  // about the initiator to decide if we can can create a session
+  message HandshakeData {
+      bytes sessionId = 1;    // for req - same as iv. for response - set to req id
+      bytes payload = 2; // empty for handshake
+      int64 timestamp = 3; // sending time
+      string clientVersion = 4; // client version of the sender
+      int32 networkID = 5; // network id of sending node
+      string protocol = 6; // 'handshake/req' || 'handshake/resp'
+      bytes nodePubKey = 7; // 65 bytes uncompressed
+      bytes iv = 8; // 16 bytes - AES-256-CBC IV
+      bytes pubKey = 9; // 65 bytes (uncompressed) ephemeral public key
+      bytes hmac = 10; // HMAC-SHA-256 32 bytes
+      string tcpAddress = 11; // ipv4 tcp address and port e.g. x.x.x.x:2424 that the remote node is accepting connections on
+      string sign = 12; // hex encoded string 32 bytes sign of all above data by node public key (verifies he has the priv key and he wrote the data
+  }
+  ```
+  - `FindNodeProtocol`  - The only `Kademlia` protocol we implement, used for bootstrapping, discovery and probing of nodes.
+```protobuf
+// FindNode is a Kademlia operation where the initiator sends a dht.ID
+// To another node and gets the list of closest nodes from the requested node.
+// If the original node isn't found by one request it keeps
+// requesting from the list until found or no new results
+message FindNodeReq {
+      Metadata metadata = 1;
+      bytes nodeId =2;
+      int32 maxResults = 3;
+}
+message FindNodeResp {
+      Metadata metadata = 1;
+      repeated NodeInfo nodeInfos = 2;
+}
+```
+
+
+  ### `Demuxer`
+  The `Demuxer` is a junction of protocols and handlers, it enables `Swarm` to take care only of validation of a message and then route to the registered protocols through `Demuxer`.
+
+  `Demuxer` exposes 2 main functions for registering and routing protocol messages.
+
+  - `RegisterProtocolHandler(handler ProtocolRegistration)` - Let's new protocols register themselves.
+  - `RouteIncomingMessage(msg IncomingMessage)` - Routes the message to the appropriate handler
+
+
+  ### Protocols
+  All protocols include `Metadata` about the author or sender of the message.
+
```

Comment:
Since the information in Metadata is p2p related it should be populated by the p2p layer instead of the protocols layer. 
IMO reqId should be implemented by each protocol and p2p shouldn't care about it